src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -258,25 +261,40 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend_reply(MPIR_Context_id_t context_i
     ucx_hdr.handler_id = handler_id;
     ucx_hdr.data_sz = data_sz;
 
-    /* just pack and send for now */
-    send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
-    MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
-    MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
     if (dt_contig) {
-        MPIR_Memcpy(send_buf + am_hdr_sz + sizeof(ucx_hdr), (char *) data + dt_true_lb, data_sz);
+        send_buf = MPL_malloc(sizeof(ucx_hdr) + am_hdr_sz, MPL_MEM_BUFFER);
+        MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
+        MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
+        iov[0].buffer = send_buf;
+        iov[0].length = sizeof(ucx_hdr) + am_hdr_sz;
+        iov[1].buffer = (char *) data + dt_true_lb;
+        iov[1].length = data_sz;
+
+        dt = ucp_dt_make_iov();
+        total_sz = 2;
     } else {
+        send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
+
+        MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
+        MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
         MPI_Aint actual_pack_bytes;
         mpi_errno =
             MPIR_Typerep_pack(data, count, datatype, 0, send_buf + am_hdr_sz + sizeof(ucx_hdr),
                               data_sz, &actual_pack_bytes);
         MPIR_ERR_CHECK(mpi_errno);
         MPIR_Assert(actual_pack_bytes == data_sz);
+
+        dt = ucp_dt_make_contig(1);
+        total_sz = data_sz + am_hdr_sz + sizeof(ucx_hdr);
     }
-    ucp_request = (MPIDI_UCX_ucp_request_t *) ucp_tag_send_nb(ep, send_buf,
-                                                              data_sz + am_hdr_sz +
-                                                              sizeof(ucx_hdr),
-                                                              ucp_dt_make_contig(1), ucx_tag,
-                                                              &MPIDI_UCX_am_isend_callback);
+    ucp_request =
+        (MPIDI_UCX_ucp_request_t *) ucp_tag_send_nb(ep,
+                                                    (dt ==
+                                                     ucp_dt_make_contig(1)) ? send_buf : (char *)
```

Comment:
cast to `char *` shouldn't be necessary, right? the argument is a `void *` so should accept any pointer type.

Reply:
I will remove it

Reply:
Following on that: the branch here is essentially the same as `if (dt_contig)`. Could this buffer setting be combined with that if/else block?

Reply:
What about a pointer `send_buf_p` that is set to pointer either to `iov` or `send_buf`?

Reply:
Yes, that would work.